### ISSUE TITLE
input validation for directory/json names

### ISSUE THREAD

**Issue by atravitz** [OpenFreeEnergy/openfe#986]
If a user provides a json file name formatted as `[non-existent directory]/filename.json`, the code runs without error, but the json file(s) are never created. This results in the user needing to rerun costly simulations.

The expected behavior is to throw an error for this up front, or create the directory on the fly

**Comment by IAlibay** [comment]
Yes, I also think it's might be possible to overwite the output JSON (I haven't tested this, this is purely based on what I've been getting as reports from other folks).

**Comment by mikemhenry** [comment]
I think we should modify the behavior slightly to this:

1. We shouldn't overwrite output jsons, and we should do this check at the start and not after a simulation is done and we are writing the results
2. I don't think we need to require the `-d directory` to be empty since even if every transformation used the same directory, the simulations shouldn't end up overwriting each other in the shared-xxx and scratch-xxx dire
3. We should create any directories we need to for the `-d` and `-o` flags if they don't exist. We should create them before we actually execute the dag, so that if we run into any issues (like permissions or something) it fails fast before we waste time running the dag and then try to save the results at the end.

As it stands, the current behavior breaks how things worked in 1.0.1. I ran into this when running some jobs from Hannah using the submission scripts from the industry benchmark since I like to programmatically create the output directories like `transformation_results_$n` where `n` is the replica number.

This was the error I got:
```
Usage: openfe quickrun [OPTIONS] TRANSFORMATION
Try 'openfe quickrun -h' for help.

Error: Invalid value for '-o': Cannot write to results_transformations_tyk2_0/rbfe_ejm_42_ejm_54.json, parent directory does not exist.
done
```

I will say the error message is excellent (thanks @atravitz ) since it clearly explained what the error was and what I needed to do. To me it feels too much like an extra hoop to jump through to add a `mkdir transformation_results_$n` before my quick run command.

**Comment by atravitz** [comment]
thanks @mikemhenry  for flagging that this breaks current behavior! 

1. This is the current behavior, unless I'm misunderstanding you. If a .json output file already exists, an error is thrown up-front.
2. We don't currently require the `-d` directory to empty, so I think this is fine as-is.
3. I agree with creating `-d` if it doesn't exist (the current behavior). However, there's a complication where I believe we don't want to create the `-o` file until it's time to write, since an empty or incomplete file would change current behavior. Both `openfe gather` and users (I think) use the existence of a file as an indicator the simulation is running or complete.

**Comment by mikemhenry** [comment]
RE: 1, I was just explaining how I thought it should work, even though that is the current behavior.

RE: 2, Same as 1.

RE: 3, Sorry I wasn't suggesting that we _make_ the file (I agree that would be new behavior and confusing), but just make sure that we can make the file, so if we need to make some directories we do. I don't think we need to go so far as touch some file to make sure we can write and then delete it (but we could using a `try:` `finally:` block)
